text/0000-keep-key-logs.md

```diff
@@ -0,0 +1,63 @@
+# RFC: Keep Key Logs
+
+- RFC PR: 
+- Tracking issue:
+
+## Motivation
+
+To address https://github.com/tikv/tikv/issues/10841, we decide to prioritize availability over log completeness thus switch the overflow strategy of logger from `Block` to `Drop`. 
+
+However, some logs are useful for diagnosis in practice. We heavily depend on logs to investigate bugs or abnormalities of the system. If all logs are dropped randomly, reasoning about the raftstore or transaction issues may be impossible.
+
+Therefore, it is only a temporary workaround to simply drop overflowed logs. We hope to keep logs that are important for diagnosis if possible.
+
+## Detailed design
+
+The problem is that we don't define which logs are important. Whether a log is important for diagnosis is not related to the log level. 
+
+For example, we prefer to keep info level logs in the raftstore rather than keep warn level logs in the coprocessor module. The former one is critical for finding out how each region evolves while the latter one may be only a retryable error.
```

Comment:
Is it possible to set different log level for different modules?

Reply:
Yes, it is. But module level setting is coarse grained. This RFC enables you to even choose which logs can be dropped in the same module.

Reply:
We used to have this problem too. Error log is not enough for diagnostics, info log is too much. 
The solution we make is to cache a buffer for current logs and as long as we find an error log, we print the whole log buffer (including every logs before the error log). 
I guess it may mitigate the problem.

Reply:
A concern here is:
In many cases we need to troubleshoot the resource-intensive scenarios and understand why it happened. 
Hence we need actually need the logs even more in such scenarios. This then leads to a paradox. 
I agree with Tony on the suggestion. 

We can create a circular buffer to keep diagnostic information and flush before a key point.

Reply:
Tony's solution sounds good but may be not so suitable for today's TiKV. TiKV is weak at detecting problems autonomously. It is really common that service downgrade or data corruption happens but TiKV itself does not know it and does not print any related error logs immediately. So the "key point" is hard to define.

Anyway, it probably hints what we can work on in the future.